### PR TITLE
Add support for HTML, Markdown and Vue component

### PR DESCRIPTION
**Pull request by gucong3000**
> Which issue, if any, is this issue related to?

#2592

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

### CODE REVIEW

**Comment by chrisvfritz** [comment]
I just tested this in a mid-sized Vue project and it works very well, except when setting the syntax (e.g. with `--syntax scss` on the CLI). The result is `CssSyntaxError` on every `.vue`, because their entire content is interpreted as SCSS, rather than only the content of the style tag being interpreted as SCSS.

**Comment by gucong3000** [comment]
> the syntax was correctly detected from the lang attribute.
That's exactly what I expected.

**Comment by chrisvfritz** [comment]
OK, just wanted to make sure this was expected behavior. 🙂

**Comment by jeddy3** [comment]
@gucong3000 Thanks, this PR is looking good.

> except when setting the syntax (e.g. with `--syntax scss` on the CLI).

We could improve the documentation to state that the `--syntax` option acts like an override, and it's unlikely you'll need it because the syntax is either inferred from the file extension or (if this PR is merged) from the `lang` attribute.

> Add support for HTML, Markdown and Vue component

As I said in https://github.com/stylelint/stylelint/pull/2855#pullrequestreview-61681920, I think adding default support for the `<style>` use case makes sense. This PR will need to add documentation for this behaviour, though.

**Comment by abdonrd** [comment]
Nice! Default support for the `<style>` tag would be great!

**Comment by gucong3000** [comment]
Part of  documentation is merged in another PR #2800

**Comment by gucong3000** [comment]
Some additions to the document was committed.
Corrections is welcome for my poor English.

File: docs/user-guide/cli.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ Linting all `.css` files except those within `docker` subfolders, using negation
 stylelint "**/*.css, !**/docker/**"
 ```
 
-In addition to `--syntax scss`, stylelint supports `--syntax less` and `--syntax sugarss` by default. If you're using one of the default syntaxes, you may not need to provide a `--syntax` option: non-standard syntaxes can be automatically inferred from the following file extensions: `.less`, `.scss`, and `.sss`.
+In addition to `--syntax scss`, stylelint supports `--syntax less` and `--syntax sugarss` by default. If you're using one of the default syntaxes, you may not need to provide a `--syntax` option: non-standard syntaxes can be automatically inferred from the following file extensions: `.less`, `.scss`, and `.sss`; for HTML-like or markdown formats, syntaxes can be inferred from attribute of `<style>` tag.
```
Comment (chrisvfritz) [inline docs/user-guide/cli.md]: Should the specific attributes be mentioned here? For example:

> for HTML-like or markdown formats, syntaxes can be inferred from a `lang` or `type` attribute on `<style>` tags (e.g. `lang="scss"`, `type="text/scss"`)

**Comment by chrisvfritz** [comment]
If there's other documentation needed, I'm happy to contribute if provided a list of what's still missing.

**Comment by jeddy3** [comment]
> If there's other documentation needed, I'm happy to contribute if provided a list of what's still missing.

Thanks for the offer. I've expanded on the documentation myself for now. Can someone cast their eyes over it, please?

**Comment by chrisvfritz** [comment]
It seems clear to me as a user. 🙂👍

**Comment by vega-wong** [comment]
@gucong3000 哈啰， 我昨天在用[vscode-stylelint](https://github.com/shinnn/vscode-stylelint)的时候出现了类似的问题，感谢🙏  @shinnn 作者给我耐心的答复。thanks！  我发现我需要自己额外写一个processor去处理我的问题：
[https://github.com/shinnn/vscode-stylelint/issues/111](https://github.com/shinnn/vscode-stylelint/issues/111)
vscode-stylelint的作者跟我说目前这个PR应用后可以解决我的需求，衷心希望这个功能能快速发布！

**Comment by jeddy3** [comment]
@gucong3000 Can [processors](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/processors.md) be improved to allow autofixing, or is this not viable? 

Am I right in thinking this PR will enable autofixing for markdown fences and `<style>` tags, but not for something like styled components (which use a processor - [stylelint-processor-styled-components](https://github.com/styled-components/stylelint-processor-styled-components))?

**Comment by gucong3000** [comment]
> Can processors be improved to allow autofixing, or is this not viable?

Make processors allow autofixing, this resulting will be enough to be used as a `PostCSS` syntax engine, just like `postcss-html`.
 
> Am I right in thinking this PR

Yes.

File: lib/__tests__/fixtures/style-tag.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="UTF-8">
+<title>Document</title>
+<style type="text/sugarss">
+/* Cannot trigger a syntax warning */
+a
```
Comment (CAYdenberg) [inline lib/__tests__/fixtures/style-tag.html]: Change comment here and in other fixtures: "should trigger warning ("block-no-empty") but is parsable - should not trigger syntax error"

**Comment by CAYdenberg** [comment]
I've tested this on a Vue codebase of mine and it seems to work as expected.

My issues are only with documentation, therefore. As I mention in my review and as @jeddy3 also pointed out, the clarification around the syntax option would help. I can't see this option actually being needed too often, so we should make it clear it's strictly optional.

I think we should remove the desciption of stylelint-processor-html if it won't be needed anymore (because functionality is provided in core): https://github.com/stylelint/stylelint/blob/master/docs/user-guide/processors.md

Also, we may need to clarify whether _other_ syntax processors can be used on style tags within HTML/Vue components. Let's say I have a Vue codebase that uses sass (indentation style) within <style> tags - can I lint it (assuming I have a sass processor). Does it matter what _order_ community processors are applied in?

File: lib/__tests__/fixtures/style-tag.markdown [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Markdown
+====
+
+```SugarSS
+/* Cannot trigger a syntax warning */
+a
```
Comment (CAYdenberg) [inline lib/__tests__/fixtures/style-tag.markdown]: Same change as in fixtures/style-tag.html

File: lib/__tests__/fixtures/style-tag.vue [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<style lang="sugarss">
+/* Cannot trigger a syntax warning */
+a
```
Comment (CAYdenberg) [inline lib/__tests__/fixtures/style-tag.vue]: Same change as in fixtures/style-tag.html

**Comment by CAYdenberg** [comment]
Although this (now) has two approvals I'm not sure I feel comfortable merging since it's pretty big and I'm new to the team. @jeddy3 when you get some time, would you like to take a final look?

**Comment by CAYdenberg** [comment]
> Also, we may need to clarify whether other syntax processors can be used on style tags within HTML/Vue components. Let's say I have a Vue codebase that uses sass (indentation style) within <style> tags - can I lint it (assuming I have a sass processor). Does it matter what order community processors are applied in?

There's still something bugging me about this in that .vue and .html are **file types**, which can themselves contain numerous **syntaxes** (at least .vue files can). I have to admit I'm not quite sure how I expect these types of edge cases to behave ... maybe this is too far into the weeds for documentation, I'd just like to understand it myself.

**Comment by jeddy3** [comment]
@CAYdenberg I have similar concerns. However, it feels like we are a bit thin-on-the-ground in terms of contributor time (myself included). So, I think I'm in favour of merging this is in as-is to give Vue users a good experience now. If we need to change the behaviour in the future we can do so with a breaking change.

I'll merge this on Sunday, but I'll leave it open for now in case anyone else wishes to comment.

**Comment by chrisvfritz** [comment]
@jeddy3 Thanks! I'm really excited to see this live. 😄 Do you know when it might make it into a release?

**Comment by jeddy3** [comment]
>  Do you know when it might make it into a release?

Hopefully, I'll have time tomorrow. However, there are a couple of PRs that need reviewing on the recommended config - https://github.com/stylelint/stylelint-config-recommended/pulls . Hopefully, a member @stylelint/core has time to look them over before then.

**Pull request merged**